# Using the GrainSizeTools script through Jupyter notebook: first steps

> IMPORTANT NOTE: This Jupyter notebook example only applies to GrainSizeTools v3.0+ Please check your script version before using this notebook. You will be able to reproduce all the results shown in this tutorial using the dataset provided with the script, the ```file data_set.txt```

## Running the script

The first step is to execute the code to get all the functionalities. Jupyter lab (or Jupyter notebooks) allows you to run any code using the following code snippet: ``%run + the Python file to run``. In this case you must set the full filepath that indicates where the file ``GrainSizeTools_script.py`` is located in your system. If the script was executed correctly you will see that all GrainSizeTools (GST) modules have been loaded correctly and a welcome message as follows:

In [8]:
%run C:/Users/marco/Documents/GitHub/GrainSizeTools/grain_size_tools/GrainSizeTools_script.py

As indicated in the welcome message, we can get a list of the main methods at any time by typing in the console:

In [9]:
get.functions_list()


List of main functions
summarize              -> get the properties of the data population
conf_interval          -> estimate a robust confidence interval using the t-distribution
calc_diffstress        -> estimate diff. stress from grain size using piezometers

plot.distribution      -> visualize the distribution of grain sizes and locate the averages
plot.qq_plot           -> test the lognormality of the dataset (q-q plot + Shapiro-Wilk test)
plot.area_weighted     -> visualize the area-weighed distribution of grain sizes
plot.normalized        -> visualize a normalized distribution of grain sizes

stereology.Saltykov    -> approximate the actual grain size distribution via the Saltykov method
stereology.calc_shape  -> approximate the lognormal shape of the actual distribution

You can get more information about the methods using the following ways:
    (1) Typing ? or ?? after the function name, e.g. ?summarize
    (2) Typing help plus the name of the function, e.g. help(summarize)

---

# Importing tabular data

For this, [Pandas](https://pandas.pydata.org/about/index.html) is the de facto standard Python library for data analysis and manipulation of table-like datasets (CSV, excel or text files among others). The library includes several tools for reading files and handling of missing data and when running the GrainSizeTools script pandas is imported as ```pd``` for its general use.

All Pandas methods to read data are all named ```pd.read_*``` where * is the file type. For example:

```python
pd.read_csv()      # read csv or txt files, default delimiter is ','
pd.read_table()    # read general delimited file, default delimiter is '\t' (TAB)
pd.read_excel()    # read excel files
pd.read_html()     # read HTML tables
# etc.
```
For other supported file types see https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html

The only mandatory argument for the reading methods is to define the path (local or URL) with the location of the file to be imported as follows.

In [10]:
# set the filepath, note that is enclosed in quotation marks
filepath = 'C:/Users/marco/Documents/GitHub/GrainSizeTools/grain_size_tools/DATA/data_set.txt'

# import the data
dataset = pd.read_table(filepath)

#display the data
dataset

Area  Circ.    Feret  FeretX  FeretY  FeretAngle  MinFeret  \
0        1   157.25  0.680   18.062  1535.0     0.5     131.634    13.500   
1        2  2059.75  0.771   62.097   753.5    16.5     165.069    46.697   
2        3  1961.50  0.842   57.871   727.0    65.0      71.878    46.923   
3        4  5428.50  0.709  114.657  1494.5    83.5      19.620    63.449   
4        5   374.00  0.699   29.262  2328.0    34.0      33.147    16.000   
...    ...      ...    ...      ...     ...     ...         ...       ...   
2656  2657   452.50  0.789   28.504  1368.0  1565.5     127.875    22.500   
2657  2658  1081.25  0.756   47.909  1349.5  1569.5     108.246    31.363   
2658  2659   513.50  0.720   32.962  1373.0  1586.0     112.286    20.496   
2659  2660   277.75  0.627   29.436  1316.0  1601.5     159.102    17.002   
2660  2661   725.00  0.748   39.437  1335.5  1615.5     129.341    28.025   

         AR  Round  Solidity  
0     1.101  0.908     0.937  
1     1.314  0.761     0.972  
2     1.139  0.878     0.972  
3     1.896  0.528     0.947  
4     1.515  0.660     0.970  
...     ...    ...       ...  
2656  1.235  0.810     0.960  
2657  1.446  0.692     0.960  
2658  1.493  0.670     0.953  
2659  1.727  0.579     0.920  
2660  1.351  0.740     0.960  

[2661 rows x 11 columns]

Some important things to note about the code snippet used above are:

- We used the ``pd.read_table()`` method to import the file. By default, this method assumes that the data to import is stored in a text file separated by tabs. Alternatively you can use the ``pd.read_csv()`` method (note that csv means comma-separated values) and set the delimiter to ``'\t'`` as follows: ``pd.read_csv(filepath, sep='\t')``.
- When calling the variable ``dataset`` it returs a visualization of the dataset imported, which is a tabular-like dataset with 2661 entries and 11 columns with different grain properties.

In Python, this type of tabular-like objects are called (Pandas) dataframes and allow a flexible and easy to use data analysis. Just for checking:

In [11]:
# show the variable type
type(dataset)

pandas.core.frame.DataFrame

> 👉 Pandas' reading methods give you a lot of control over how a file is read. To keep things simple, I list here the most commonly used arguments:

```python
sep         # Delimiter/separator to use.
header      # Row number(s) to use as the column names. By default it takes the first row as the column names (header=0). If there is no columns names in the file you must set header=None
skiprows    # Number of lines to skip at the start of the file (an integer).
na_filter   # Detect missing value markers. False by default.
sheet_name  # Only excel files, the excel sheet name either a number or the full name.
```

> more details on Pandas csv read method: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html

The GrainSizeTools script includes a method named ```get_filepath()``` to get the path of a file through a file selection dialog instead of directly writing it. This can be used in two ways:

```python
# store the path in a variable (here named filepath for convenience) and then use it when calling the read method
filepath = get_filepath()
dataset = pd.read_csv(filepath, sep='\t')

# use get_filepath() directly within the read method
dataset = pd.read_csv(get_filepath(), sep='\t')
```

Lastly, Pandas also allows to directly import tabular data from the clipboard (i.e. data copied using copy-paste commands). For example, after copying the table from a text/excel file or a website using: 

```python
dataset = pd.read_clipboard()
```

---

## Basic tabular data (Pandas) manipulation

